source/extensions/http/header_validators/envoy_default/header_validator.cc

```diff
@@ -8,38 +14,405 @@ namespace EnvoyDefault {
+  // Verify that the header name is valid. This also honors the underscore in
+  // header configuration setting.
+  //
+  // From RFC 9110, https://www.rfc-editor.org/rfc/rfc9110.html#section-5.1:
+  //
+  // header-field   = field-name ":" OWS field-value OWS
+  // field-name     = token
+  // token          = 1*tchar
+  //
+  // tchar          = "!" / "#" / "$" / "%" / "&" / "'" / "*"
+  //                / "+" / "-" / "." / "^" / "_" / "`" / "|" / "~"
+  //                / DIGIT / ALPHA
+  //                ; any VCHAR, except delimiters
+  const auto& key_string_view = name.getStringView();
+  bool allow_underscores =
+      config_.headers_with_underscores_action() == HeaderValidatorConfig::ALLOW;
+  // This header name is initially invalid if the name is empty.
+  if (key_string_view.empty()) {
+    return {HeaderEntryValidationResult::Action::Reject,
+            UhvResponseCodeDetail::get().EmptyHeaderName};
+  }
+
+  bool is_valid = true;
+  char c = '\0';
+
+  for (auto iter = key_string_view.begin(); iter != key_string_view.end() && is_valid; ++iter) {
+    c = *iter;
+    is_valid &= testChar(kGenericHeaderNameCharTable, c) && (c != '_' || allow_underscores);
+  }
+
+  if (!is_valid && c == '_' &&
+      config_.headers_with_underscores_action() == HeaderValidatorConfig::DROP_HEADER) {
+    return {HeaderEntryValidationResult::Action::DropHeader,
+            UhvResponseCodeDetail::get().InvalidUnderscore};
+  } else if (!is_valid) {
+    auto details = c == '_' ? UhvResponseCodeDetail::get().InvalidUnderscore
+                            : UhvResponseCodeDetail::get().InvalidCharacters;
+    return {HeaderEntryValidationResult::Action::Reject, details};
+  }
+
   return HeaderEntryValidationResult::success();
 }
 
-::Envoy::Http::HeaderValidator::RequestHeaderMapValidationResult
-HeaderValidator::validateRequestHeaderMap(::Envoy::Http::RequestHeaderMap&) {
-  return RequestHeaderMapValidationResult::success();
+HeaderValidator::HeaderValueValidationResult
+HeaderValidator::validateGenericHeaderValue(const HeaderString& value) {
+  // Verify that the header value is valid.
+  //
+  // From RFC 9110, https://www.rfc-editor.org/rfc/rfc9110.html#section-5.5:
+  //
+  // header-field   = field-name ":" OWS field-value OWS
+  // field-value    = *field-content
+  // field-content  = field-vchar
+  //                  [ 1*( SP / HTAB / field-vchar ) field-vchar ]
+  // field-vchar    = VCHAR / obs-text
+  // obs-text       = %x80-FF
+  //
+  // VCHAR          =  %x21-7E
+  //                   ; visible (printing) characters
+  const auto& value_string_view = value.getStringView();
+  bool is_valid = true;
+
+  for (auto iter = value_string_view.begin(); iter != value_string_view.end() && is_valid; ++iter) {
+    is_valid &= testChar(kGenericHeaderValueCharTable, *iter);
+  }
+
+  if (!is_valid) {
+    return {HeaderValueValidationResult::Action::Reject,
+            UhvResponseCodeDetail::get().InvalidCharacters};
```

Comment:
optional: should we have different details for invalid keys vs invalid values?  I think it'd be nice.